### PR TITLE
created the block details page

### LINKED ISSUE #7: Block details page
# Block details page

## ⚠️ Reading contributors guidelines to get assigned is MANDATORY!

[Read contributors guidelines](https://github.com/walnuthq/op-scan/issues/1)

## User stories

As a user, when I access a block details page, I want to see basic info about the block such as block number, block hash, mining timestamp, number of transactions contained in the block as well as total gas used by the block.
When I click on previous block / next block buttons I want to navigate between adjacent blocks.
When I click on the number of transactions I want to be redirected to the block transactions list: `/block/:number/txs`.

## Validation

It should look like the Etherscan block detail page, eg. https://optimistic.etherscan.io/block/121511346
Do NOT use an accordion to show more details, just display block hash below block number.
Do NOT implement the block transactions list page.

## Implementation

You can use the transaction details page as inspiration for the global architecture.
The URL of the page should be `/block/:number`.
You should use a description list similar to the tx details page to display the block attributes and the pagination component for the previous / next block buttons.

## Resources

- tx details page: https://github.com/walnuthq/op-scan/blob/main/src/components/pages/tx/index.tsx
- shadcn/ui pagination: https://ui.shadcn.com/docs/components/pagination

### PR DESCRIPTION
**Pull request by Dprof-in-tech**
closes #7 

Here is what the block details page looks like 

image

### CODE REVIEW

File: pnpm-lock.yaml
Comment (saimeunt) [inline pnpm-lock.yaml:1]: Don't commit your pnpm-lock.yaml as it's been generated with an outdated pnpm version.
Checkout the version from main instead.

File: src/app/block/[block]/page.tsx [hunk 1]
```diff
@@ -1,5 +1,105 @@
-const BlockPage = ({ params: { block } }: { params: { block: string } }) => (
-  <h1>{block}</h1>
-);
+import { createPublicClient, http, formatEther } from 'viem';
+import { optimism } from 'viem/chains';
+import Link from 'next/link';
+import { ChevronLeft, ChevronRight } from 'lucide-react';
 
-export default BlockPage;
+const client = createPublicClient({
```
Comment (saimeunt) [inline src/app/block/[block]/page.tsx]: We already define a `l2PublicClient` in `@/lib/chains.ts`, use this instead.

File: src/app/block/[block]/page.tsx [hunk 2]
```diff
@@ -1,5 +1,105 @@
...
+  chain: optimism,
+  transport: http(process.env.RPC_URL),
+});
+
+function formatTimestamp(timestamp: number): string {
```
Comment (saimeunt) [inline src/app/block/[block]/page.tsx]: Use the `formatTimestamp` from `@/lib/utils.ts`.

File: src/app/block/[block]/page.tsx [hunk 3]
```diff
@@ -1,5 +1,105 @@
...
+    const formattedTimestamp = formatTimestamp(Number(blockData.timestamp));
+
+    return (
+      <div className="container mx-auto px-4 py-8">
+        <h1 className="text-lg font-semibold mb-4 mt-0">Block #{blockNumber.toString()}</h1>
+        <div className="shadow overflow-hidden sm:rounded-lg">
+          <div className="border-t border-gray-200 px-4 py-5 sm:p-0">
+            <dl className="sm:divide-y sm:divide-gray-200">
+            <div className="py-4 sm:py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
+                <dt className="text-sm font-medium ">Block Height:  </dt>
```
Comment (saimeunt) [inline src/app/block/[block]/page.tsx]: Use `DescriptionListItem` from `@/components/lib/description-list-item.tsx` to ensure consistency between pages.

File: src/app/block/[block]/page.tsx [hunk 4]
```diff
@@ -1,5 +1,105 @@
...
+              </div>
+              <div className="py-4 sm:py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
+                <dt className="text-sm font-medium ">Gas Used</dt>
+                <dd className="mt-1 text-sm  sm:mt-0 sm:col-span-2">{blockData.gasUsed.toString()}</dd>
+              </div>
+              <div className="py-4 sm:py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
+                <dt className="text-sm font-medium ">Gas Limit</dt>
+                <dd className="mt-1 text-sm sm:mt-0 sm:col-span-2">{blockData.gasLimit.toString()}</dd>
+              </div>
+              <div className="py-4 sm:py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
```
Comment (saimeunt) [inline src/app/block/[block]/page.tsx]: Remove baseFeePerGas item as it's not deemed necessary by Etherscan.

File: src/app/block/[block]/page.tsx [hunk 5]
```diff
@@ -1,5 +1,105 @@
...
+                <dt className="text-sm font-medium ">Transactions</dt>
+                <dd className="mt-1 text-sm  sm:mt-0 sm:col-span-2">
+                  <Link href={`/block/${blockNumber}/txs`} className="text-blue-600 hover:underline">
+                    {blockData.transactions.length} transactions
+                  </Link>
+                </dd>
+              </div>
+              <div className="py-4 sm:py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
+                <dt className="text-sm font-medium ">Gas Used</dt>
+                <dd className="mt-1 text-sm  sm:mt-0 sm:col-span-2">{blockData.gasUsed.toString()}</dd>
```
Comment (saimeunt) [inline src/app/block/[block]/page.tsx]: Format `gasUsed` using a NumberFormatter and en-US locale so we have commas.
See some examples here: https://github.com/walnuthq/op-scan/blob/main/src/lib/utils.ts

Also, we need the percentage of gas used over gas limit displayed in parenthesis.

File: src/app/block/[block]/page.tsx [hunk 6]
```diff
@@ -1,5 +1,105 @@
...
+              </div>
+              <div className="py-4 sm:py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
+                <dt className="text-sm font-medium ">Gas Limit</dt>
+                <dd className="mt-1 text-sm sm:mt-0 sm:col-span-2">{blockData.gasLimit.toString()}</dd>
```
Comment (saimeunt) [inline src/app/block/[block]/page.tsx]: Again, this should be formatted using NumberFormatter.

**Review by saimeunt — CHANGES_REQUESTED** [review]
@Dprof-in-tech please carefully review and make the requested changes to get this merged.

**Comment by Dprof-in-tech** [comment]
all changes has been made as requested
image

**Pull request closed without merging**